### ISSUE TITLE
Add way to not fetch data twice in `load` when provided `fetch` cannot be used

### ISSUE THREAD

**Issue by dodiameer** [sveltejs/kit#3729]
### Describe the problem

A lot of services provide SDKs used to simplify fetching data from their APIs, however a decent amount of them don't provide an option for a custom `fetch` function, so when you use that SDK in `load()` the data will be fetched twice, both on the server and the client. A library I'm having trouble with is Ghost CMS's SDK, and I remember having this issue with something else too but I can't remember what it was. 

### Describe the proposed solution

A function passed to load could allow you to pass in a unique ID and a callback — it runs that callback, caches the result as JSON and returns it. Once the app loads on the client-side and runs load again this function could check the cache and return it. For example:

```ts
export async function load({ cache, params }) {
  const post = cache(`post_${params.slug}`, async () => {
    try {
      return await someSdk.getPost(params.slug)
    } catch {
      return null
    }
  })
  // Do other things in load
  return { status: 200, props: { post } }
}
```

By caching I mean store the JSON string in the document the same way the custom fetch does - not sure if cache is the right term here 

### Alternatives considered

Using endpoints that run the SDK functions, but it's a bit boilerplate and loses some type-safety

### Importance

nice to have

**Comment by Rich-Harris** [comment]
The API you describe makes sense. I'm also not sure what the best name is, but will throw out `serde` as an option since we're talking about serializing and deserializing (not sure how widely known that contraction is, but it'd be easy to search for in the docs at least).

The cache key would be unnecessary in most cases — it could default to an incrementing integer. (If you started doing things inside `if` blocks and `for` loops then you'd need it, of course.)

So it could look like this:

```js
export async function load({ serde, params }) {
  const post = await serde(() => someSdk.getPost(params.slug).catch(() => null));

  // Do other things in load
  return { status: 200, props: { post } }
}
```

## Alternatives

I've honestly yet to find an SDK that wouldn't be better as a vanilla REST API — the more we can encourage platforms to offer endpoints that can be used with vanilla `fetch` (or at _least_ allow a custom `fetch` implementation to be passed in, so that they can be used across different environments without bundling `node-fetch` etc), the better. 

Assuming that's not possible, a good option is to proxy the API request. Now that we have [shadow endpoints](https://github.com/sveltejs/kit/pull/3679), this is extremely easy:

```js
// src/routes/blog/[slug].js
export function get({ params }) {
  try {
    const post = await someSdk.getPost(params.slug).catch(() => null);
    return { body: { post } };
  } catch {
    return { status: 404 };
  }
}
```

```svelte
<script>
  <!-- look ma, no `load`! -->
  export let post;
</script>
```

The downside is that you add a network hop, of course. Depending on how you've deployed your app that may or may not be an issue.

**Comment by synthXR** [comment]
Would be awesome to implement. Our app loads data twice on every page load (ideally should be loaded on the server, and cached for the client). Or there could be an option to avoid re-loading data by the client if the server already loaded it.

**Comment by kenfoo** [comment]
Are there any future plans for this?

In my case, what I have is a server that holds a snapshot of some latest data from an API endpoint. 
This API endpoint is able to provide new data incrementally in small bits, provided I have a snapshot of the current data.

What I'd like to do is have the server render the html page with the latest data that it has, whilst handing over the snapshot
of the latest data to the client as a prop so that the client may store it and request for incremental bits of data on its own from time to time.

On the server side this is fine as load() is able to access the server's persistent store of its latest data and build the data incrementally.

On the client side however, the same load() function runs with no initial state when the page is initially loaded, and has to resort to fetching the full dataset from the API endpoint. (Instead of being able to build it incrementally with an initial set of data provided by the server during SSR).

Granted, this happens only once on the initial page load, but for a high traffic website, performing the initial fetch of the full (possibly large) dataset could be taxing for the API endpoint.

I'm rather new to svelte/sveltekit, so with my limited experience at the moment I don't really see a way around this.
(I've seen some stackoverflow workarounds about sending the initial state as a cookie, but it sounds wrong to me.)

**Comment by jdgamble555** [comment]
@kenfoo - The current _svelte way_ of doing this would be to use an endpoint and an observable.  Get your data in an endpoint, put it in a load function with fetch, and render the page as normal.  The endpoint should only be called once, as that is the point of the fetch.  Then, take that initial state value that you returned from the load function, and insert that into an observable in the `onMount()` function to incrementally update the page however you like.  In the end, you have the server call, and the observable call, but that is it.

I hope this helps.

J

FWI - Rich Harris has talked about maybe one day doing something like [NextJS's server side streaming](https://nextjs.org/docs/api-reference/next/streaming), but that is probably a long way away and does not support your current use case.

**Comment by kenfoo** [comment]
Thanks @jdgamble555   I just found out about the new shadow endpoints and that seems to solve my use case perfectly.

**Comment by kpietraszko** [comment]
Apart from the above mentioned Firebase and Apollo, it's also important for gRPC (or tRPC for that matter). 
Using endpoints is not really a solution, that extra app complexity and network hop is unnecessary. 
I feel SvelteKit should probably expose a way to inline the data in the HTML, like fetch does, manually.

**Comment by Rich-Harris** [comment]
We've been thinking further about what a sensible API for this could look like in #9160, so I'm going to close this issue in favour of that one
